test/Dialect/Torch/maximize-value-semantics.mlir

```diff
@@ -59,6 +59,17 @@ func @mutation_followed_by_view_like_ops(%value_t: !torch.vtensor, %overwriter:
   return %value_result : !torch.vtensor
 }
 
+// CHECK-LABEL:   func @mutation_of_view_like_op_result(
+// CHECK-SAME:                                             %[[VALUE_T:.*]]: !torch.vtensor, %[[OVERWRITER:.*]]: !torch.vtensor, %[[INT_LIST:.*]]: !torch.list<!torch.int>) -> !torch.vtensor {
+// CHECK:           return %[[OVERWRITER]] : !torch.vtensor
+func @mutation_of_view_like_op_result(%value_t: !torch.vtensor, %overwriter: !torch.vtensor, %int_list: !torch.list<!torch.int>) -> !torch.vtensor {
+  %t = torch.copy.to_tensor %value_t : !torch.tensor
+  %view = torch.aten.view %t, %int_list : !torch.tensor, !torch.list<!torch.int> -> !torch.tensor
+  torch.overwrite.tensor.contents %overwriter overwrites %view : !torch.vtensor, !torch.tensor
+  %result = torch.copy.to_vtensor %view : !torch.vtensor
```

Comment:
I think we also need to make sure there is not any use of `%t` after the `torch.overwrite.tensor.contents`.  The problem with overwriting the result of a view like op is that the result is aliased with the input operands of view like ops ((or other ops with aliasing). 

I am not sure how can we model the case where there is indeed use of `%t` after the overwrite but maybe that's not common in real models? @silvasean Do you have any comments?

Reply:
That's a really good point. Yeah if there's a use of `%t` after the overwrite, that would complicate the analysis. At the moment I can't think of a way of doing it. However, verifying that the intermediate values of view-like ops between the `torch.copy.to_tensor` and the `torch.overwrite.tensor.contents` are not used after the overwrite should be straight forward.

Reply:
I am thinking we can handle both view like ops and overwrites in the same pattern. What I have in mind is combining `AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock` and `RewriteViewLikeSubgraph` into one conversion pattern. 

First step is collecting all the aliasing non-value tensors similar to this part https://github.com/llvm/torch-mlir/blob/486f95e84f587d020ba789b071b12f890510f1a1/lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp#L130-L141. In addition to collecting the users as the existing code, all the non-value tensors need to be collected as well. Let's call this aliasing set for now.

The second step is doing abstract interpretation similarly to `AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock`.  All the collected users in the first step needs to be sorted. At each operation, only one value from the aliasing set can be available (meaning its current value can be replaced with a value tensor). At the `CopyToNonValueTensorOp`, its result non value tensor is available with value of its operand. Whenever, a `overwrite` is seen, the overwritten non value tensor become available and others in the aliasing set are unavailable. If any uses of an unavailable non value tensor is seen the pattern fails.  If the interpretation finished successfully, all the overwrites can be removed. All uses of the aliasing set should be replaced with corresponding available value tensor.

Reply:
Yi, that seems like a good generalization of the current logic. One thing to keep in mind is the handling of "read-only" ops in nested subregions (like in If/Loop). From the perspective of this algorithm, they can be treated as if they were in the parent block. If there are mutating ops in subregions, then we cannot abstractly interpret past them (without doing something a lot more sophisticated AFAICT).